plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java

```diff
@@ -196,49 +195,81 @@ private CollectedStatistics mergeStatisticsIfNecessary(
-                                Integer fieldId = getOnlyElement(read.first().inputFields());
-                                checkState(pendingPreviousNdvSketches.remove(fieldId), "Unwanted read of stats for field %s", fieldId);
-                                Memory memory = Memory.wrap(ByteBuffers.getBytes(read.second())); // Memory.wrap(ByteBuffer) results in a different deserialized state
-                                CompactSketch previousSketch = CompactSketch.wrap(memory);
-                                CompactSketch newSketch = requireNonNull(collectedNdvSketches.get(fieldId), "ndvSketches.get(fieldId) is null");
-                                ndvSketches.put(fieldId, SetOperation.builder().buildUnion().union(previousSketch, newSketch));
+            if (hasUsefulData) {
+                try (PuffinReader reader = Puffin.read(fileIO.newInputFile(statisticsFile.path()))
+                        .withFileSize(statisticsFile.fileSizeInBytes())
+                        .withFooterSize(statisticsFile.fileFooterSizeInBytes())
+                        .build()) {
+                    List<BlobMetadata> toRead = reader.fileMetadata().blobs().stream()
+                            .filter(blobMetadata -> blobMetadata.type().equals(APACHE_DATASKETCHES_THETA_V1))
+                            .filter(blobMetadata -> blobMetadata.inputFields().size() == 1)
+                            .collect(toImmutableList());
+                    for (Pair<BlobMetadata, ByteBuffer> read : reader.readAll(toRead)) {
+                        Integer fieldId = getOnlyElement(read.first().inputFields());
+                        if (pendingPreviousNdvSketches.remove(fieldId)) {
+                            CompactSketch newSketch = requireNonNull(collectedNdvSketches.get(fieldId), "ndvSketches.get(fieldId) is null");
+                            if (shouldMergeWithOld) {
+                                newSketch = SetOperation.builder().buildUnion().union(readPreviousSketch(read), newSketch);
                             }
+                            ndvSketches.put(fieldId, newSketch);
                         }
-                        catch (IOException exception) {
-                            throw new UncheckedIOException(exception);
+                        else {
+                            ndvSketches.put(fieldId, readPreviousSketch(read));
                         }
                     }
                 }
-
-                yield new CollectedStatistics(ndvSketches.buildOrThrow());
+                catch (IOException exception) {
+                    throw new UncheckedIOException(exception);
+                }
+                pendingPreviousNdvSketches.forEach(newFieldId -> ndvSketches.put(newFieldId, collectedNdvSketches.get(newFieldId)));
             }
-        };
+            return (Map<Integer, CompactSketch>) ndvSketches.buildOrThrow();
+        }).orElse(collectedNdvSketches);
```

Comment:
` }).orElse(collectedNdvSketches);` does not look correct when `shouldMergeWithOld` is true

Reply:
i think this was naming problem

Reply:
please explain

Reply:
I meant that this may seems incorrect because of the name `shouldMergeWithOld` I choosed for that variable, now after some refactoring it is gone.
My assumption is that if we get to `mergeStatisticsIfNecessary` but there is no previous stats existing we should just take stats that we calculated during this call. This `orElse(collectedNdvSketches)` code will be called only in such a situation.